File: leetcode-java/src/completefifties/l0000/ReverseNodesInKGroup.java
package completefifties.l0000;

import tools.ListNode;

public class ReverseNodesInKGroup {

    /**
     * 25. Reverse Nodes in k-Group
     *
     * @return
     * k = 2: 2->1->4->3->5
     * k = 3: 3->2->1->4->5
     */
    private ListNode reverseKGroup(ListNode head, int k) {
        ListNode prev = null;
        ListNode cur = head;
        ListNode next = null;
        ListNode check = head; // why check can't be declared before cur???
        int canProceed = 0;
        int count = 0;

        // check if the list can be reversed
        while (canProceed < k && check != null) {
            check = check.next;
            canProceed++;
        }

        // The list can be reversed
        if (canProceed == k) {
            while (count < k && cur != null) {
                next = cur.next;
                cur.next = prev;
                prev = cur;
                cur = next;
                count++;
            }
            if (next != null) {
                // head is the reversed list's tail
                head.next = reverseKGroup(next, k);
            }
            // prev is the reversed list's head
            return prev;
        } else {
            // The list can't be reversed, directly return the list itself
            return head;
        }
    }
}

File: leetcode-python/linked_list/reverse_nodes_in_k_group.py
from tools.node import ListNode


def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
    """
    Stack
    112ms 21.14%
    13.9MB 99.63%
    :param self:
    :param head:
    :param k:
    :return:
    """
    dummy = ListNode(0)
    p = dummy

    while True:
        count = k
        stack = []
        tmp = head

        while count and tmp:
            stack.append(tmp)
            tmp = tmp.next
            count -= 1

        # 注意,目前tmp所在k+1位置
        # 说明剩下的链表不够k个,跳出循环
        if count:
            p.next = head
            break

        # 翻转操作
        while stack:
            p.next = stack.pop()
            p = p.next

        # 与剩下链表连接起来
        p.next = tmp
        head = tmp

    return dummy.next


def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
    """
    尾插法
    :param self:
    :param head:
    :param k:
    :return:
    """
    dummy = ListNode(0)
    dummy.next = head
    pre = dummy
    tail = dummy

    while True:
        count = k
        while count and tail:
            count -= 1
            tail = tail.next

        if not tail:
            break

        head = pre.next

        while pre.next != tail:
            cur = pre.next
            pre.next = cur.next
            cur.next = tail.next
            tail.next = cur

        pre = head
        tail = head

    return dummy.next


def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
    """
    递归
    :param self:
    :param head:
    :param k:
    :return:
    """
    cur = head
    count = 0
    while cur and count!= k:
        cur = cur.next
        count += 1
    if count == k:
        cur = self.reverseKGroup(cur, k)
        while count:
            tmp = head.next
            head.next = cur
            cur = head
            head = tmp
            count -= 1
        head = cur
    return head
